docs/plus/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,192 @@
+# HorizontalPodAutoscaler
+
+HorizontalPodAutoscaler allows your services to scale up when demand is high and scale down when they are no longer needed.
+
+!!! tip ""
+    [API Reference](../reference/cdk8s-plus-24/typescript.md#horizontalpodautoscaler)
+
+## Using a HorizontalPodAutoscaler
+
+The example below creates a HorizontalPodAutoscaler that scales the number of replicas of a Deployment based on the average CPU utilization of the pods.
+
+```typescript
+import * as k from 'cdk8s';
+import * as kplus from 'cdk8s-plus-25';
+
+const app = new k.App();
+const chart = new k.Chart(app, 'Chart');
+
+// Lets create a deployment for our web server
+const bookstoreWebsite = new kplus.Deployment(chart, 'BookstoreWebsite', {
+  containers: [ { image: 'node' } ],
+});
+
+// Now define a HorizontalPodAutoscaler
+const hpa = new kplus.HorizontalPodAutoscaler(chart, 'BookstoreWebsiteHpa', {
+  target: bookstoreWebsite,
+  maxReplicas: 10,
+  metrics: [
+     kplus.Metric.resourceCpu(kplus.MetricTarget.averageUtilization(70)),
+  ],
+ });
+// This will scale our website deployment up when the average CPU utilization
+// is 70% or higher up to a maximum of 10 replicas.
+```
+
+Using the example above, if our web server passes 70% CPU utilization the autoscaler will try and figure out the number of replicas to scale.
+
+The autoscaler uses the following formula:
+
+> desiredReplicas = ceil[currentReplicas * ( currentMetricValue / desiredMetricValue )]
+
+Based on this formula we can find out how many replicas will be scaled. For example if we have 1 replica and the CPU utilization is at 70%.
+
+> desiredReplicas = ceil[1 * ( 70 / 70 )] = 1
+
+The autoscaler will try and add 1 replica, meaning we'll have 2 web servers running, the original and the new replica.
+
+If the CPU utilization is 140% the autoscaler will try and provision 2 replicas. Scaling us up to 3 web servers total.
+
+> desiredReplicas = ceil[1 * ( 140 / 70 )] = 2
+
+This works the same way when scaling down.
+
+For example if our target CPU utilization is 70% and we have 3 replicas running and the CPU utilization is at 50%.
+
+> desiredReplicas = ceil[3 * ( 50 / 70 )] = 2
+
+The autoscaler will try and remove 1 replica, meaning we'll have 2 web servers running, the original and the new replica.
+
+
+## Metrics
+
+The HorizontalPodAutoscaler supports the following metrics:
+
+### Resource Metrics
+
+Resource metrics are used to scale on a resource like CPU or memory.
+
+```typescript
+import * as kplus from 'cdk8s-plus-25';
+
+
+const hpa = new kplus.HorizontalPodAutoscaler(chart, 'BookstoreWebsiteHpa', {
+  target: bookstoreWebsite,
+  maxReplicas: 10,
+  metrics: [
+     kplus.Metric.resourceCpu(kplus.MetricTarget.averageUtilization(70)),
+  ],
+ });
+```
+
```

Comment:
We should mention the note [k8s docs](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-resource-metrics) provides for resource metrics:

> Note: Since the resource usages of all the containers are summed up the total pod utilization may not accurately represent the individual container resource usage. This could lead to situations where a single container might be running with high usage and the HPA will not scale out because the overall pod usage is still within acceptable limits.